### PR TITLE
Tests for API, more input checks

### PR DESCRIPTION
**Pull request by tommyod**
- Using `ALLOWED_DISTRS` instead of re-defining allowed distributions in the context of testing.
- Added more sanitation of user inputs to `fit` and `score`.
- Wrote some tests to solidify the API.

### CODE REVIEW

File: pyglmnet/pyglmnet.py [hunk 1]
```diff
@@ -692,8 +695,20 @@ def fit(self, X, y):
...
+            raise ValueError('Input data should be of shape' +
+                             '(n_observations, n_features)')
+
+        y = np.asarray(y)
+        if y.ndim == 2:
+            y = y.squeeze()
+
+        n_observations, n_features = X.shape
 
+        if not n_observations == len(y):
```
Comment (jasmainak) [inline pyglmnet/pyglmnet.py]: if n_observations != len(y)

File: pyglmnet/pyglmnet.py [hunk 2]
```diff
@@ -692,8 +695,20 @@ def fit(self, X, y):
             raise ValueError('Input data should be of type ndarray (got %s).'
                              % type(X))
 
-        n_features = X.shape[1]
+        if not X.ndim == 2:
```
Comment (jasmainak) [inline pyglmnet/pyglmnet.py]: if X.ndim != 2

File: pyglmnet/pyglmnet.py [hunk 3]
```diff
@@ -692,8 +695,20 @@ def fit(self, X, y):
...
+            raise ValueError('Input data should be of shape' +
+                             '(n_observations, n_features)')
+
+        y = np.asarray(y)
+        if y.ndim == 2:
```
Comment (jasmainak) [inline pyglmnet/pyglmnet.py]: why not raise an error for y the same way as for X?

File: tests/test_pyglmnet.py [hunk 4]
```diff
@@ -388,3 +385,61 @@ def test_random_state_consistency():
 
     assert_array_equal(ypred_a, ypred_b)
     assert_array_equal(ypred_b, ypred_c)
+
+
+@pytest.mark.parametrize("distr", ALLOWED_DISTRS)
+def test_simulate_glm(distr):
+    """
+    Test that every generative model can be simulated from.
+    """
```
Comment (jasmainak) [inline tests/test_pyglmnet.py]: numpy doctring conventions dictate that L392-L394 should be in one line

File: tests/test_pyglmnet.py [hunk 5]
```diff
@@ -388,3 +385,61 @@ def test_random_state_consistency():
...
+
+    n_samples, n_features = 10, 3
+
+    # sample random coefficients
+    beta0 = np.random.normal(0.0, 1.0, 1)
+    beta = np.random.normal(0.0, 1.0, n_features)
+
+    np.random.seed(1)
+    X = np.random.normal(0.0, 1.0, [n_samples, n_features])
+    simulate_glm(distr, beta0, beta, X)
```
Comment (jasmainak) [inline tests/test_pyglmnet.py]: this is just a smoke test. Is there a benefit to having it separately than what already exists?

File: tests/test_pyglmnet.py [hunk 6]
```diff
@@ -388,3 +385,61 @@ def test_random_state_consistency():
...
+
+
+def test_simulate_glm_failing():
+    """
+    Test that an error is raised if an invalid parameter name is passed.
+    """
```
Comment (jasmainak) [inline tests/test_pyglmnet.py]: same comment as above, should be in one line

File: tests/test_pyglmnet.py [hunk 7]
```diff
@@ -388,3 +385,61 @@ def test_random_state_consistency():
...
+    distr = 'multivariate_gaussian_poisson'
+    with pytest.raises(ValueError):
+        simulate_glm(distr, 1.0, 1.0, np.array([[1.0]]))
```
Comment (jasmainak) [inline tests/test_pyglmnet.py]: my rule of thumb is that a function should be at least 10 lines, otherwise it shouldn't be a function

File: tests/test_pyglmnet.py [hunk 8]
```diff
@@ -388,3 +385,61 @@ def test_random_state_consistency():
...
+
+
+def test_API_shape_mismatch():
+    """
+    Test that a ValueError is raised if the shapes mismatch.
+    """
+    X = np.random.randn(10, 5)
+    y = np.random.randn(11)
+    with pytest.raises(ValueError):
+        GLM().fit(X, y)
```
Comment (jasmainak) [inline tests/test_pyglmnet.py]: same comment here. These functions seem to be a little fragmented. I'd recommend pooling them into one

File: tests/test_pyglmnet.py [hunk 9]
```diff
@@ -388,3 +385,61 @@ def test_random_state_consistency():
...
+    y = y_func(y)
+
+    glm = GLM(distr='gaussian')
+    glm.fit(X, y)
+
+    glm.predict(X)
+    glm.score(X, y)
+
+
+def test_API_shape_mismatch():
```
Comment (jasmainak) [inline tests/test_pyglmnet.py]: it's not kosher to have caps in a function name

File: tests/test_pyglmnet.py [hunk 10]
```diff
@@ -388,3 +385,61 @@ def test_random_state_consistency():
...
+    X = np.random.randn(n_samples, n_features)
+    y = np.random.randn(n_samples)
+
+    y = y_func(y)
+
+    glm = GLM(distr='gaussian')
+    glm.fit(X, y)
+
+    glm.predict(X)
+    glm.score(X, y)
```
Comment (jasmainak) [inline tests/test_pyglmnet.py]: just a smoke test. I would suggest checking if errors are raised when it's not the right dimensions

File: pyglmnet/pyglmnet.py (hunk 3 above)
Comment (tommyod) [reply pyglmnet/pyglmnet.py]: As it stands the code allows both y to be of shape (obs,) or (obs, 1).

File: tests/test_pyglmnet.py (hunk 4 above)
Comment (tommyod) [reply tests/test_pyglmnet.py]: Done.

File: tests/test_pyglmnet.py (hunk 6 above)
Comment (tommyod) [reply tests/test_pyglmnet.py]: Done

**Comment by tommyod** [comment]
Thank you for the prompt review! I've addressed some of the comments in a commit, and would like to discuss the rest.

The main reason why I added the smoke-tests was because I was figuring out the API, and I wanted to solidify it in the code somewhere. It was hard to get started with the code, since:
- The code in the `master` README does not run with code from the `master`.
- There are no short examples in docstrings showing usage.
- There's an `examples` folder, but the examples are very long and contain dependencies, i.e. `from spykes.ml.strf import STRF`.

Adding API usage as a that might have been the wrong approach. Instead, I propose:
- Adding 1-2 short examples showing usage as doctests in the GLM object, and running these on travis CI.
- Adding a few short examples in `examples` and also running these on travis CI.

Running examples on travis CI is important, if not they might diverge from the code in the future. Thoughts on the approach sketched above?

**Comment by jasmainak** [comment]
> The code in the master README does not run with code from the master.

PR to fix this would be welcome :)

> There are no short examples in docstrings showing usage.

again, PR is welcome!

> There's an examples folder, but the examples are very long and contain dependencies, i.e. from spykes.ml.strf import STRF

yes, I know it's a little unfortunate. I am also not sure know how much spykes is maintained. Do you know other data that could be used to demonstrate the same functionality? Tikhonov regularization in this case ...

> Running examples on travis CI is important, if not they might diverge from the code in the future. 

Currently we do have CircleCI for examples. I'm not quite sure why it's not getting triggered on PRs.

**Comment by jasmainak** [comment]
okay I _think_ I fixed the CircleCI issue. If you make a push, it should run CircleCI

**Comment by tommyod** [comment]
@jasmainak , I've:

- Updated the README example, it runs now.
- Added a tiny doctest. More can be added later.
- Merged some of the small tests.

I believe this PR is good to go. CircleCI is failing, seemingly because `plot_group_lasso.py` takes too long. Seems unrelated to this PR.

File: pyglmnet/pyglmnet.py [hunk 11]
```diff
@@ -448,6 +451,23 @@ class GLM(BaseEstimator):
...
         default: False
 
+    Examples
+    --------
+    >>> import numpy as np
+    >>> random_state = 1
+    >>> np.random.seed(random_state)
+    >>> n_samples, n_features = 100, 4
+    >>> X = np.random.randn(n_samples, n_features)
+    >>> y = y = 2.2 * X[:, 0] -1.0 * X[:, 1] + 0.3 * X[:, 3] + 1.0
```
Comment (jasmainak) [inline pyglmnet/pyglmnet.py]: y = 2.2 ...

instead of 

y = y = 2.2

File: tests/test_pyglmnet.py [hunk 12]
```diff
@@ -388,3 +385,47 @@ def test_random_state_consistency():
...
+def test_simulate_glm(distr):
+    """Test that every generative model can be simulated from."""
+
+    n_samples, n_features = 10, 3
+
+    # sample random coefficients
+    beta0 = np.random.normal(0.0, 1.0, 1)
+    beta = np.random.normal(0.0, 1.0, n_features)
+
+    np.random.seed(1)
```
Comment (jasmainak) [inline tests/test_pyglmnet.py]: why not use random_state in simulate_glm? It's the preferred approach

File: pyglmnet/pyglmnet.py (hunk 3 above)
Comment (jasmainak) [reply pyglmnet/pyglmnet.py]: hmm, that's confusing indeed. I think historically it was allowed because we had a multinomial distribution but now that's removed it shouldn't be the case.

File: tests/test_pyglmnet.py [hunk 13]
```diff
@@ -388,3 +385,47 @@ def test_random_state_consistency():
...
+
+    # If the distribution name is garbage it will fail
+    distr = 'multivariate_gaussian_poisson'
+    with pytest.raises(ValueError):
+        simulate_glm(distr, 1.0, 1.0, np.array([[1.0]]))
+
+
+@pytest.mark.parametrize("y_func",
+                         [list, tuple, np.array,
+                          lambda x: np.array(x).reshape(-1, 1)])
```
Comment (jasmainak) [inline tests/test_pyglmnet.py]: see comment above. Probably better to remove this case of y as a 2D array. I think it could lead to weird corner cases that we should avoid

File: tests/test_pyglmnet.py [hunk 14]
```diff
@@ -388,3 +385,47 @@ def test_random_state_consistency():
...
+def test_api_input_types_y(y_func):
+    """Test that the input value of y can be of different types."""
+    np.random.seed(1)
```
Comment (jasmainak) [inline tests/test_pyglmnet.py]: better to use random state where possible. It does not do global seed resetting

File: pyglmnet/pyglmnet.py [hunk 15]
```diff
@@ -448,6 +451,23 @@ class GLM(BaseEstimator):
...
+    >>> glm = GLM(distr='gaussian', verbose=False, random_state=random_state)
+    >>> glm = glm.fit(X, y)
+    >>> glm.beta0_ # The intercept
+    1.005377890812967
+    >>> glm.beta_ # The coefficients
+    array([ 1.90217526, -0.78781579, -0.        ,  0.03227754])
```
Comment (jasmainak) [inline pyglmnet/pyglmnet.py]: sorry my brain is fried at this moment. So I apologize if this is a very dumb question. But can you explain why these are not the same as the coefficients above?

File: pyglmnet/pyglmnet.py [hunk 16]
```diff
@@ -448,6 +451,23 @@ class GLM(BaseEstimator):
     verbose : boolean or int
         default: False
 
+    Examples
+    --------
+    >>> import numpy as np
+    >>> random_state = 1
+    >>> np.random.seed(random_state)
```
Comment (jasmainak) [inline pyglmnet/pyglmnet.py]: as commented elsewhere, this is not a recommended approach. It leads to reproducibility issues

File: pyglmnet/pyglmnet.py (hunk 3 above)
Comment (tommyod) [reply pyglmnet/pyglmnet.py]: Only allowing 1D arrays for y now, and validating against this.

File: tests/test_pyglmnet.py (hunk 12 above)
Comment (tommyod) [reply tests/test_pyglmnet.py]: Changed it now. Can you please review how I do it and check if it's correct?

File: tests/test_pyglmnet.py (hunk 13 above)
Comment (tommyod) [reply tests/test_pyglmnet.py]: Changed this test, only allowing ndarrays for both X and y now.

**Comment by tommyod** [comment]
- Changed input to `fit` to only allow ndarrays. Simplified validation and the tests.
- Used RandomState instead of `np.random.seed`. Can you check if it's correct now?
- Regarding the comment about the coefficients after fitting not being equal to those in the input- can it be due to regularization and the optimization routine(e.g. the number of iterations run and the tolerance)? The difference in the coefficients is kind of large though.

File: pyglmnet/pyglmnet.py (hunk 15 above)
Comment (tommyod) [reply pyglmnet/pyglmnet.py]: See the comment below. I'm not entirely sure - it could be the regularization or the optimization routine. They seem to be quite far off though.

Comment (jasmainak) [reply pyglmnet/pyglmnet.py]: hmm ... I wonder if this is related to #275 or #278. Does this also happen when you set the reg to 0?

**Comment by jasmainak** [comment]
Merged @tommyod . Thanks!

Let's iterate on the issue with the coefficient in another PR -- it seems unrelated to the changes in this PR.

I had to squash because you had a merge commit in your history. Next time rebase instead of merging or pulling to avoid that.

File: README.rst [hunk 17]
```diff
@@ -119,42 +119,43 @@ Install ``pyglmnet`` using ``setup.py`` as follows
 Getting Started
 ~~~~~~~~~~~~~~~
 
+
 Here is an example on how to use the ``GLM`` estimator.
+.. This example is also found in examples/intro_example.py.
```
Comment (jasmainak) [inline README.rst]: @tommyod, it passed my attention when reviewing. Looks like you forgot to commit this file
